### ISSUE TITLE
Add docs hint to use getNumProcessors for determining shakeThreads automatically

### ISSUE THREAD

**Issue by nh2** [ndmitchell/shake#221]
I often use

```haskell
main :: IO ()
main = do
  nprocs <- getNumProcessors

  shakeArgs shakeOptions { shakeThreads = nprocs } $ ...
```

to automatically determine the number of processors.

This might be nice to mention in the docs, especially because `getNumProcessors` can easily be confused with `getNumCapabilities` which is usually not what one wants to use for `shakeThreads`.

---

However, I got tricked by combining this with the comment `Omitting -threaded will still allow your cmd actions to run in parallel` [in the GHC BUILD FLAGS section of the docs](https://hackage.haskell.org/package/shake-0.14.3/docs/Development-Shake.html):

Not setting `-threaded` will make `getNumProcessors` return `1` always, so by using `getNumProcessors` AND the above advice, I get exactly the opposite of what I intended ;)

So if we add a hint to `getNumProcessors`, we should also mention that in that case the user should use `-threaded`.

**Comment by ndmitchell** [comment]
I have an internal version of `getNumProcessors` in Shake, called [`getProcessorCount`](https://github.com/ndmitchell/shake/blob/9b3962df17d4d2d90e5d41e4548e0b15d1f9949a/src/General/Extra.hs#L44), which does the right thing in most cases even without `-threaded`. You can trigger than behaviour by setting `shakeThreads` to 0, something which I documented only yesterday: https://github.com/ndmitchell/shake/commit/2ff5a21d252cf32f2151d68b6a8d30a825d32bb9

Does that work for you? Might even simplify your code.

**Comment by nh2** [comment]
Yes that would work, and indeed be very comfortable.

How does that work on Windows though? It seems to me that there's nothing that stops that function from doing `readFile "/proc/cpuinfo"`.

**Comment by ndmitchell** [comment]
The behavior of 0 being guess at the processors has been available since 0.11, so you can rely on it already, it's just I forgot to document it until  yesterday. On Windows it will try reading `/proc/cpuinfo`, but there's a `handle` that means if it throws an error (which it probably will) we just default to 1.

**Comment by nh2** [comment]
> but there's a handle that means if it throws an error (which it probably will) we just default to 1.

Ah right, when skimming it I somehow got confused thinking that `handle_` is a file `Handle` instead of being a `Control.Exception` function ;)

> The behavior of 0 being guess at the processors has been available since 0.11

Ah great, tried it, works well for me!

--

What isn't clear to me yet is whether Shake recommends linking with `-threaded` for GHC 7.8 upwards, or not? The manual says for <= 7.6, better not use `-threaded` due to that bug, but what's preferred for GHCs that don't have that bug?

**Comment by ndmitchell** [comment]
I've updated the docs to make it clear that the advice to omit `-threaded` is only 7.6 and below. I'm a bit wary of advising something that isn't a good idea on certain GHC's (it makes for an easy mistake to make), but once GHC 7.10 is out it I'll start to update the defaults.
